### ISSUE TITLE
State Legislature / Executive

### ISSUE THREAD

**Issue by dpeters911** [5calls/5calls#167]
This is great for national issues but right now the placeholders are limited to Federal.

Some of the most pressing issues are done at the state level for now.

As far as I know, Google Civics is missing placeholder contacts for these state level legislators (had trouble pulling them up on Google Civics API demo page).

Any chance in future?

**Comment by dpeters911** [comment]
Even allowing for zip lookup based on airtable data would be great, then we could handle gaps in Google Civics on our own and get granular.

**Comment by ctaymor** [comment]
It seems very hit or miss whether Google Civic's API has state legislators. They had mine for my home address, but for a number of addresses of public locations around here, they did not. Perhaps they only have that info for residential addresses.

I tried a few other random addresses (selected by zooming into suburban areas in Google maps in different cities until I came up with a single house address) and they did have state legislature info.

It seems like it might be worth doing even if we could only start out by covering the addresses with data in the Google Civic's API.

**Comment by jaymcgrath** [comment]
This API could be useful for local legislators: 
https://sunlightfoundation.com/api/
docs: http://docs.openstates.org/en/latest/api/

**Comment by nickoneill** [comment]
Yeah, local issues are definitely something we're looking into how to support. Obviously a lot more work to generate issues for more local areas!

Thanks for the tips on APIs as well!

**Comment by jlev** [comment]
OpenStates is the place to go for this info, in my opinion. It's used by other big civic action apps (eg, ActionNetwork), and the reasons why a state is or is not covered are public, unlike Google Civic.

**Comment by nickoneill** [comment]
Very interesting @jlev, I like the idea that they're more open about what is and is not covered (and being able to open an issue with broken stuff is nice). I'll take a look but we would definitely want governors in there since we already have (and use) that info from google civic before we swapped over.

**Comment by dpeters911** [comment]
Alternative:
Let us define placeholders groups using a string in a column, I.e. "LocalMayor" for infinitely scabable placeholder groups. We can use that for everything from school boards to even non-government organization structures.

At SeattleDSA, we've been organizing around hyperlocal issues or contacting committees. The data is available but is manually assigned one-to-many in AirTables per issue when being able to define a committee or group would be useful.

**Comment by nickoneill** [comment]
@dpeters911 can you explain more or link to a read-only version of your airtable? We've been using placeholders (US SENATE, US HOUSE, GOVERNOR - you know) for civic data but I'm not sure how we could utilize this for data not already coming back from another source.

**Comment by dpeters911** [comment]
Here is an example. I'm hovering over the add column (+). Let's say this column is called PlaceholderAlias

![image](https://cloud.githubusercontent.com/assets/5984719/23010622/531ccd7e-f3d1-11e6-81c8-9f3c5a675c15.png)

A PlaceholderAlias would work like LOCAL REP does. It would bring up a group of contacts.

However, instead of pulling those contacts from a third-party API like Google Civic, it would simply reference other contacts listed. Contacts could be in many PlaceholderAliases. In the example above, many of the state senators are in a Committee. This is a great use for a PlaceholderAlias, esp. when bills haven't reached the floor for a vote. It also allows us to be nimble for unions, special task forces, caucuses, etc. that arise ad-hoc.

This also for infinite scaling to self-managed contacts outside Google Civics API. We can adapt it locally, easily creating groups on the fly.

**Comment by oderby** [comment]
I think improving functionality around aliases is a separate problem to fix (and one that I think we can definitely improve greatly, either as @dpeters911 suggests or otherwise), that probably should be filed and tracked in another issue.

@nickoneill perhaps I'm misunderstanding, but I don't think we have to "swap over" wholesale to sunlight or other apis - we should be able to "just" layer/merge the responses from both apis (and any others we want to use)? While there's a lot of technical difficulties and corner-cases involved, conceptually it's pretty simple - when looking up reps for an address, query all available apis and merge (union) the results into a single response.

**Comment by agjohnson** [comment]
+1 for layering in, I'm not sure we'd have comparable results with the openstates/sunlight geolookup. It could be worth trying however.

Looking at the openstates API, there will likely be a few snags:

* Nomenclature and existence of lower/upper legislative bodies varies state by state. I'm not sure if this is important, but this data exists in the openstate state API if that is important/required: http://docs.openstates.org/en/latest/api/metadata.html#id3
* Openstates doesn't return executive positions, not sure if there is a good source for this yet

Otherwise, lookup against the openstates api will use civic OCD division information, so at least this part will be mostly straight forward. It might warrant adding a separate cache layer here, so caching happens against the civic lookup by address/zip/location, and caching also happens per division. This would reduce lookups a bit more, as division will be more likely to get a cache hit.

If we did want to supplement with Airtable data, using the OCD division id, or at least a key using this information, might be better than using a zip code for lookup.

If no one has delved into this by this weekend, I'll likely give the backend pieces for this a go.

**Comment by nickoneill** [comment]
@oderby @agjohnson True, we don't have to stop using civic to get other specific data from different sources. That might be the best plan.

I'm also interested in a platform like geocod.io that has both state and national districts baked in to a geocoding service. Then we could either use the civic API for rep info or (more likely) provide our own info based on the district info returned.

**Comment by oderby** [comment]
> I'm also interested in a platform like geocod.io that has both state and national districts baked in to a geocoding service.

Agreed - that'd be great. But also... isn't that what civic api is **supposed** to do? Do we have any insight into why the data from Google is unreliable? What I'm getting at is do we have any reason to expect such a service exists, if Google can't even get it working right? And even if such a service did exist, do we expect it to be more accurate/reliable than civic api? (It's possible civic api was just a side-project of some googler that is no longer being maintained, and so the answer to these questions could be 'yes'). Or can we at least hope it'll fail in different ways/on different inputs (i.e. failures are uncorrelated, which is a requirement for any result fusing approach to work). My point is that we may end up finding such a layering approach may be required anyways in the end.

**Comment by nickoneill** [comment]
@oderby @agjohnson Check out https://github.com/CivilServiceUSA/api as well. Long chat with the guy who runs this today and it's incredibly well put together (data right now, but will be in the API soon). Could replace the civic api for both geocoding *and* data, including more local stuff than we can get from civic.

**Comment by nickoneill** [comment]
(I totally agree that we could do a better job of quantifying what's wrong with the civic API but critically the fact that the data is all hidden away and we have to wait for them to update it after x election is a big failing for me)

**Comment by oderby** [comment]
> Check out https://github.com/CivilServiceUSA/api as well. Long chat with the guy who runs this today and it's incredibly well put together (data right now, but will be in the API soon). Could replace the civic api for both geocoding and data, including more local stuff than we can get from civic.

Hmm, sounds great but won't be a viable solution until it's implemented and running. Did he give an idea of his timeline? Regardless, looks like a great project to keep track of for now.

> critically the fact that the data is all hidden away and we have to wait for them to update it after x election is a big failing for me

Agreed. I would love to see us move in a direction where we're not so tightly coupled to civic api.

**Comment by oderby** [comment]
Just came across this https://www.270towin.com/elected-officials/ - seems like they're using proprietary data purchased from http://greatdata.com/. Dropping here just for future reference.

**Comment by piebob** [comment]
just noting a user request (from email) for this feature. 
if we eventually do some kind of priority feature/work planning, i'd like a way to track how many users requested a thing, etc. if there's a better way than comments like this one, please suggest :)

**Comment by nickoneill** [comment]
This is probably the right way for now @piebob, thanks!
